crates/ra_assists/src/auto_import.rs

```diff
@@ -0,0 +1,780 @@
+    pub fn leading_indent(node: &SyntaxNode) -> Option<&str> {
+        for leaf in prev_leaves(node) {
+            if let Some(ws) = ast::Whitespace::cast(leaf) {
+                let ws_text = ws.text();
+                if let Some(pos) = ws_text.rfind('\n') {
+                    return Some(&ws_text[pos + 1..]);
+                }
+            }
+            if leaf.leaf_text().unwrap().contains('\n') {
+                break;
+            }
+        }
+        None
+    }
+
+    fn prev_leaves(node: &SyntaxNode) -> impl Iterator<Item = &SyntaxNode> {
+        generate(prev_leaf(node), |&node| prev_leaf(node))
+    }
+
+    fn prev_leaf(node: &SyntaxNode) -> Option<&SyntaxNode> {
+        generate(node.ancestors().find_map(SyntaxNode::prev_sibling), |it| {
+            it.last_child()
+        })
+        .last()
+    }
+}
+
+fn collect_path_segments(path: &ast::Path) -> Option<Vec<&ast::PathSegment>> {
+    let mut v = Vec::new();
+    collect_path_segments_raw(&mut v, path)?;
+    return Some(v);
+}
+
+fn collect_path_segments_raw<'b, 'a: 'b>(
+    segments: &'b mut Vec<&'a ast::PathSegment>,
+    mut path: &'a ast::Path,
+) -> Option<usize> {
+    let oldlen = segments.len();
+    loop {
+        let mut children = path.syntax().children();
+        let (first, second, third) = (
+            children.next().map(|n| (n, n.kind())),
+            children.next().map(|n| (n, n.kind())),
+            children.next().map(|n| (n, n.kind())),
+        );
+        match (first, second, third) {
+            (Some((subpath, PATH)), Some((_, COLONCOLON)), Some((segment, PATH_SEGMENT))) => {
+                path = ast::Path::cast(subpath)?;
+                segments.push(ast::PathSegment::cast(segment)?);
+            }
+            (Some((segment, PATH_SEGMENT)), _, _) => {
+                segments.push(ast::PathSegment::cast(segment)?);
+                break;
+            }
+            (_, _, _) => return None,
+        }
+    }
+    // We need to reverse only the new added segments
+    let only_new_segments = segments.split_at_mut(oldlen).1;
+    only_new_segments.reverse();
+    return Some(segments.len() - oldlen);
+}
+
+fn fmt_segments(segments: &[&ast::PathSegment]) -> String {
+    let mut buf = String::new();
+    fmt_segments_raw(segments, &mut buf);
+    return buf;
+}
+
+fn fmt_segments_raw(segments: &[&ast::PathSegment], buf: &mut String) {
+    let mut first = true;
+    for s in segments {
+        if !first {
+            buf.push_str("::");
+        }
+        match s.kind() {
+            Some(ast::PathSegmentKind::Name(nameref)) => buf.push_str(nameref.text()),
+            Some(ast::PathSegmentKind::SelfKw) => buf.push_str("self"),
+            Some(ast::PathSegmentKind::SuperKw) => buf.push_str("super"),
+            Some(ast::PathSegmentKind::CrateKw) => buf.push_str("crate"),
+            None => {}
+        }
+        first = false;
+    }
+}
+
+#[derive(Copy, Clone)]
+enum PathSegmentsMatch {
```

Comment:
Hm, I wonder if instead of `enum` we could return a single `usize` -- length of the common prefix. It sort-of has the same amount of info as this `enum`, and doesn't suffer the problem that `Empy` and `Partial(0)` are the same.